samples: bluetooth: mesh: sensor sample features

HaavardRei opened the pull request:
Updates the sensor sample with a range limit setting property
for the chip temperature sensor.

Adds a descriptor to the chip temperature sensor.

The client sample sends:
A get sensor range setting message on button 1 press.
A set sensor range setting message on button 2 press.
A get sensor descriptor message on button 3 press.

Adds functionality for persistently storing the range limit setting.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei commented on subsys/bluetooth/mesh/shell/shell_sensor_cli.c:
```diff
@@ -275,7 +275,13 @@ static int setting_set(const struct shell *shell, size_t argc, char *argv[], boo
 	uint32_t sensor_id = shell_strtoul(argv[1], 0, &err);
 	uint32_t setting_id = shell_strtoul(argv[2], 0, &err);
-	struct sensor_value value = shell_model_strtosensorval(argv[3], &err);
+	const struct bt_mesh_sensor_type *sensor_type = bt_mesh_sensor_type_get(sensor_id);
```
Changes in this file is very hacky and is done for testing purposes.

Andrewpini commented:
![image](https://user-images.githubusercontent.com/35562897/188577287-ed016b50-a280-45d4-8736-bf5d027be185.png)
Looks like you could benefit from a round of `Crtl+k Ctrl+f`

Andrewpini commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -36,56 +33,112 @@ static void sensor_cli_data_cb(struct bt_mesh_sensor_cli *cli,
+		err = bt_mesh_sensor_cli_setting_set(&sensor_cli, NULL,
+						     &bt_mesh_sensor_present_dev_op_temp,
+						     &bt_mesh_sensor_dev_op_temp_range_spec,
+						     &temp_ranges[tidx++ % ARRAY_SIZE(temp_ranges)],
```
> C:/Git_repos/ncs/nrf/samples/bluetooth/mesh/sensor_client/src/model_handler.c: In function 'button_handler_cb':
C:/Git_repos/ncs/nrf/samples/bluetooth/mesh/sensor_client/src/model_handler.c:130:12: warning: passing argument 5 of 'bt_mesh_sensor_cli_setting_set' from incompatible pointer type [-Wincompatible-pointer-types]    
  130 |            &temp_ranges[tidx++ % ARRAY_SIZE(temp_ranges)],
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |            |
      |            const struct bt_mesh_sensor_column *
In file included from C:/Git_repos/ncs/nrf/include/bluetooth/mesh/models.h:41,
                 from C:/Git_repos/ncs/nrf/samples/bluetooth/mesh/sensor_client/src/model_handler.c:8:
C:/Git_repos/ncs/nrf/include/bluetooth/mesh/sensor_cli.h:504:29: note: expected 'const struct sensor_value *' but argument is of type 'const struct bt_mesh_sensor_column *'
  504 |  const struct sensor_value *value,

Im not sure this is completly correct

Andrewpini replied:
I have investigated this a bit. You could try using a multidimentional array. What about something like this?:
```
static const struct sensor_value temp_ranges[][2] = {
	{{ 0, 0 }, { 100, 0 }},
	{{ 10, 0 }, { 20, 0 }},
	{{ 22, 0 }, { 30, 0 }},
	{{ 40, 0 }, { 50, 0 }},
};

static void button_handler_cb(uint32_t pressed, uint32_t changed)
{
               ...

		err = bt_mesh_sensor_cli_setting_set(&sensor_cli, NULL,
						     &bt_mesh_sensor_present_dev_op_temp,
						     &bt_mesh_sensor_dev_op_temp_range_spec,
						     &temp_ranges[tidx++ % ARRAY_SIZE(temp_ranges)][0],
						     NULL);

               ...
}
```

Andrewpini replied:
Casting the sensor column could work to, but im not sure if that is the "correct" way to handle it:
```
		err = bt_mesh_sensor_cli_setting_set(
			&sensor_cli, NULL, &bt_mesh_sensor_present_dev_op_temp,
			&bt_mesh_sensor_dev_op_temp_range_spec,
			(const struct sensor_value *)&temp_ranges[tidx++ % ARRAY_SIZE(temp_ranges)],
			NULL);
```

Andrewpini commented on samples/bluetooth/mesh/sensor_server/src/model_handler.c:
```diff
@@ -65,21 +79,100 @@ static int chip_temp_get(struct bt_mesh_sensor_srv *srv,
 static struct bt_mesh_sensor chip_temp = {
 	.type = &bt_mesh_sensor_present_dev_op_temp,
 	.get = chip_temp_get,
+	.descriptor = &chip_temp_descriptor,
+	.settings = {
+		.list = &chip_temp_setting,
+		.count = ARRAY_SIZE(chip_temp_setting),
+	},
```
```suggestion
	.descriptor = &chip_temp_descriptor,
	.settings = {
		.list = (const struct bt_mesh_sensor_setting *)&chip_temp_setting,
		.count = ARRAY_SIZE(chip_temp_setting),
	},
```
Fixes compile warning

HaavardRei commented on samples/bluetooth/mesh/sensor_server/src/model_handler.c:
```diff
@@ -65,21 +79,100 @@ static int chip_temp_get(struct bt_mesh_sensor_srv *srv,
 	return err;
 }
+/* Tolerance is based on nRF52832's range and accuracy (5/125 = 4%). */
+static const struct bt_mesh_sensor_descriptor chip_temp_descriptor = {
```
The descriptor has been added, but doesn't really do anything on its own. I was able to print the info using the shell commands (descriptor get). Could for instance make one of the buttons trigger sending the message instead?

Andrewpini replied:
Ye, sure you could try that

Andrewpini commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -36,56 +33,112 @@ static void sensor_cli_data_cb(struct bt_mesh_sensor_cli *cli,
 static struct k_work_delayable get_data_work;
 
 static void get_data(struct k_work *work)
 {
```
It would be nice if you could prevent attempts to send messages as long as the device is not provisioned. See `bt_mesh_is_provisioned()`. An example of how it can be used is present in the light_switch sample.

Andrewpini commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -36,56 +33,112 @@ static void sensor_cli_data_cb(struct bt_mesh_sensor_cli *cli,
+}
 
-	req_temp_range = !req_temp_range;
+static uint32_t tidx;
```
If this variable is only used inside the button handler, it should be decleared there. Also I would prefer if you changed the variable name to something else, like `temp_idx`. `tidx` can be a bit ambiguous since `tid` often will be read as "Transaction ID"

Andrewpini commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -36,56 +33,112 @@ static void sensor_cli_data_cb(struct bt_mesh_sensor_cli *cli,
+					 const struct bt_mesh_sensor_type *sensor,
+					 const struct bt_mesh_sensor_setting_status *setting)
+{
+	printk("Sensor ID: 0x%04x, ", sensor->id);
+	printk("Setting ID: 0x%04x\n", setting->type->id);
```
```suggestion
	printk("Sensor ID: 0x%04x, Setting ID: 0x%04x\n", sensor->id, setting->type->id);
```

Andrewpini commented on samples/bluetooth/mesh/sensor_client/README.rst:
```diff
@@ -34,7 +34,8 @@ Overview
 The following Bluetooth mesh sensor types are used in this sample:
 
-* :c:var:`bt_mesh_sensor_present_dev_op_temp` - Published by the server according to its publishing period.
+* :c:var:`bt_mesh_sensor_present_dev_op_temp` - Published by the server according to its publishing period. <--- Is it?
```
```suggestion
* :c:var:`bt_mesh_sensor_present_dev_op_temp` - Published by the server according to its publishing period. <--- It does ;)?
```

Andrewpini commented on samples/bluetooth/mesh/sensor_client/README.rst:
```diff
@@ -80,6 +81,12 @@ Buttons:
+Button 1:
+   Sends a setting get message (after the provisioning procedure is finished).
+
+Button 2:
```
Could be worth mentioning that it will toggle between the different ranges.

HaavardRei replied:
I have :)

Andrewpini commented on samples/bluetooth/mesh/sensor_server/README.rst:
```diff
@@ -82,7 +83,10 @@ The models are used for the following purposes:
 The model handling is implemented in :file:`src/model_handler.c`.
 It uses the TEMP_NRF5 or BME680 temperature sensor depending on the platform.
-The :ref:dk_buttons_and_leds_readme library is used to detect button presses.
+The sample has a descriptor related to the :c:var:`bt_mesh_sensor_present_dev_op_temp` sensor, which specifies tolerance values for the ``TEMP_NRF5`` temperature sensor calculated based on the ``nRF52832 Product Specification <--- Link somehow``.
```
You could ask Mia Koen if she knows a way to link the specification.

HaavardRei replied:
Awaiting an answer!

Andrewpini reviewed (changes requested):
In general I think this looks good 👍 Make sure that all compliznce errors are resolved and take a look at my comments. While reviewing your code I found a bug that caused the Client application to Hard Fault during provisioning, but this was not something that has been introduced by you. I have created a fix for this bug that is in review ATM, so when it is ready you can rebase on top of the main branch again.

HaavardRei commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -36,56 +33,142 @@ static void sensor_cli_data_cb(struct bt_mesh_sensor_cli *cli,
+	       bt_mesh_sensor_ch_str(&sensor->descriptor.tolerance.positive),
+	       bt_mesh_sensor_ch_str(&sensor->descriptor.tolerance.negative));
+	printk("\tsampling type: %d\n", sensor->descriptor.sampling_type);
+	printk("\tperiod: %lld\n", sensor->descriptor.period);
```
The period and update interval is not defined in the descriptor (the descriptor type mentions these as optional), perhaps remove these prints.

HaavardRei commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -36,56 +33,142 @@ static void sensor_cli_data_cb(struct bt_mesh_sensor_cli *cli,
+	static uint32_t temp_idx;
+
+	if ((pressed & BIT(0))) {
+		int err;
```
Should the err declaration be moved to the outside of the if-s?

Andrewpini replied:
Yes, you can make one common instance for all buttons 👍

Andrewpini commented:
After resolving the last few things you mentioned I think the PR is ready for the "Real" review 👍

Andrewpini reviewed (approved):
👍

HaavardRei committed:
samples: bluetooth: mesh: sensor sample features

Updates the sensor sample with a range limit setting property
for the chip temperature sensor.

Adds a descriptor to the chip temperature sensor.

The client sample sends:
A get sensor range setting message on button 1 press.
A set sensor range setting message on button 2 press.
A get sensor descriptor message on button 3 press.

Adds functionality for persistently storing the range limit setting.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

omkar3141 commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -42,13 +42,37 @@ static void sensor_cli_series_entry_cb(
+	printk("Descriptor of sensor with ID 0x%04x:\n", sensor->id);
+	printk("\ttolerance: { positive: %s",
+	       bt_mesh_sensor_ch_str(&sensor->descriptor.tolerance.positive));
+	printk(" negative: %s }\n", bt_mesh_sensor_ch_str(&sensor->descriptor.tolerance.negative));
```
```suggestion
	printk("\t negative: %s }\n", bt_mesh_sensor_ch_str(&sensor->descriptor.tolerance.negative));
```
Should this be formatted correctly?

omkar3141 commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -58,11 +82,25 @@ static struct k_work_delayable get_data_work;
+		return;
+	}
+
+	static uint32_t sensor_idx;
```
```suggestion
	static uint32_t sensor_idx = 0;
```

omkar3141 commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -71,21 +109,68 @@ static void get_data(struct k_work *work)
+static const struct sensor_value temp_ranges[][2] = {
+	{ { 0 }, { 100 } },
+	{ { 10 }, { 20 } },
+	{ { 22 }, { 30 } },
```
Should it be ...
```suggestion
	{ { 20 }, { 30 } },
```
?

omkar3141 commented on samples/bluetooth/mesh/sensor_client/src/model_handler.c:
```diff
@@ -71,21 +109,68 @@ static void get_data(struct k_work *work)
+		}
+	}
+	if (pressed & BIT(1)) {
+		err = bt_mesh_sensor_cli_setting_set(
```
It will be nice to print what range is being set.

omkar3141 commented on samples/bluetooth/mesh/sensor_server/README.rst:
```diff
@@ -82,8 +83,14 @@ The models are used for the following purposes:
 The model handling is implemented in :file:`src/model_handler.c`.
 It uses the TEMP_NRF5 or BME680 temperature sensor depending on the platform.
+
+The sample has a descriptor related to the :c:var:`bt_mesh_sensor_present_dev_op_temp` sensor, which specifies tolerance values for the ``TEMP_NRF5`` temperature sensor calculated based on the `nRF52832 Temperature Sensor Electrical Specification`_.
```
```suggestion
The sample has a descriptor related to the :c:var:`bt_mesh_sensor_present_dev_op_temp` sensor, which specifies range setting values for the ``TEMP_NRF5`` temperature sensor. The value of the temperature reported by the sensor is calculated based on the `nRF52832 Temperature Sensor Electrical Specification`_.
```
?
